scene/2d/box2d_world.cpp

```diff
@@ -268,97 +486,71 @@ void Box2DWorld::PreSolve(b2Contact *contact, const b2Manifold *oldManifold) {
-
-				if (c_ptr->fixture_a->body_node->is_contact_monitor_enabled()) {
-					// TODO lock/unlock
-					c_ptr->fixture_a->body_node->contact_monitor->contacts.erase(*c_ptr);
-				}
-				if (c_ptr->fixture_b->body_node->is_contact_monitor_enabled()) {
-					// TODO lock/unlock
-					c_ptr->fixture_b->body_node->contact_monitor->contacts.erase(*c_ptr);
-				}
-
-				buffer_manifold->remove(i);
-				if (buffer_manifold->count == 0) {
-					ERR_FAIL_COND(!contact_buffer.erase(reinterpret_cast<uint64_t>(contact)));
-				}
-			}
-		}
-	}
-	for (int i = 0; i < b2_maxManifoldPoints; ++i) {
-		if (state2[i] == b2PointState::b2_addState) {
-			try_buffer_contact(contact, i);
-		}
-	}
-
 	if (buffer_manifold) {
-		// Only handle the first PreSolve for this contact this step (don't overwrite initial impact_velocity, world_pos)
-		for (int i = 0; i < buffer_manifold->count; ++i) {
-			Box2DContactPoint *c_ptr = &buffer_manifold->points[i];
-
-			if (c_ptr->solves == 0) {
-				c_ptr->solves += 1;
-			}
-		}
-	}
+		for (int i = b2_maxManifoldPoints - 1; i >= 0; --i) {
+			if (state1[i] == b2PointState::b2_removeState) {
+				// Remove this contact
 
-	// Handle removed/added points within the manifold
-	for (int i = b2_maxManifoldPoints - 1; i >= 0; --i) {
-		if (state1[i] == b2PointState::b2_removeState) {
-			// Remove this contact
-			if (buffer_manifold && i < buffer_manifold->count) {
 				Box2DContactPoint *c_ptr = &buffer_manifold->points[i];
 
-				if (c_ptr->fixture_a->body_node->is_contact_monitor_enabled()) {
+				if (c_ptr->id == -1)
+					continue;
+
+				if (c_ptr->fixture_a->owner_node->_is_contact_monitor_enabled()) {
 					// TODO lock/unlock
-					c_ptr->fixture_a->body_node->contact_monitor->contacts.erase(*c_ptr);
+					c_ptr->fixture_a->owner_node->contact_monitor->contacts.erase(*c_ptr);
 				}
-				if (c_ptr->fixture_b->body_node->is_contact_monitor_enabled()) {
+				if (c_ptr->fixture_b->owner_node->_is_contact_monitor_enabled()) {
 					// TODO lock/unlock
-					c_ptr->fixture_b->body_node->contact_monitor->contacts.erase(*c_ptr);
+					c_ptr->fixture_b->owner_node->contact_monitor->contacts.erase(*c_ptr);
 				}
 
-				buffer_manifold->remove(i);
-				if (buffer_manifold->count == 0) {
-					ERR_FAIL_COND(!contact_buffer.erase(reinterpret_cast<uint64_t>(contact)));
-				}
+				buffer_manifold->erase(i);
 			}
 		}
 	}
 	for (int i = 0; i < b2_maxManifoldPoints; ++i) {
 		if (state2[i] == b2PointState::b2_addState) {
-			try_buffer_contact(contact, i);
+			buffer_manifold = try_buffer_contact(contact, i);
 		}
 	}
+	if (buffer_manifold && buffer_manifold->is_empty()) {
+		ERR_FAIL_COND(!contact_buffer.erase(reinterpret_cast<uint64_t>(contact)));
+	}
 
 	if (buffer_manifold) {
 		// Only handle the first PreSolve for this contact this step (don't overwrite initial impact_velocity, world_pos)
-		for (int i = 0; i < buffer_manifold->count; ++i) {
+		for (int i = 0; i < b2_maxManifoldPoints; ++i) {
```

Comment:
I believe it's because the b2 contact can have the idx=0 manifold point be invalid while idx=1 is a valid point. Not sure the details, but the changes to the contact monitor are important bug fixes.